packages/field-base/src/slot-styles-mixin.js

```diff
@@ -0,0 +1,79 @@
+import { dedupingMixin } from '@polymer/polymer/lib/utils/mixin.js';
+
+const stylesMap = new WeakMap();
+
+/**
+ * Get all the styles inserted into root.
+ * @param {DocumentOrShadowRoot} root
+ * @return {Array<string>}
+ */
+function getRootStyles(root) {
+  if (!stylesMap.has(root)) {
+    stylesMap.set(root, {});
+  }
+
+  return stylesMap.get(root);
+}
+
+/**
+ * Insert styles into the root.
+ * @param {string} styles
+ * @param {DocumentOrShadowRoot} root
+ */
+function insertStyles(styles, root) {
+  const style = document.createElement('style');
+  style.textContent = styles;
+
+  if (root === document) {
+    document.head.appendChild(style);
+  } else {
+    root.insertBefore(style, root.firstChild);
+  }
+}
+
+const SlotStylesMixinImplementation = (superclass) =>
+  class SlotStylesMixinClass extends superclass {
+    /**
+     * List of styles to insert into root.
+     * @protected
+     */
+    get slotStyles() {
+      return {};
+    }
+
+    /** @protected */
+    connectedCallback() {
+      super.connectedCallback();
+
+      this.__applySlotStyles();
+    }
+
+    /** @private */
+    __gatherSlotStyles() {
+      if (!this.__slotStyles) {
+        this.__slotStyles = Object.entries(this.slotStyles);
+      }
+
+      return this.__slotStyles;
+    }
+
+    /** @private */
+    __applySlotStyles() {
+      const root = this.getRootNode();
+      const rootStyles = getRootStyles(root);
+      const slotStyles = this.__gatherSlotStyles();
+
+      slotStyles.forEach(([id, styles]) => {
```

Comment:
Should we namespace the `id`? Considering the following case:

`<web-component-one>`:
```js
get slotStyles() {
  return {
    prefix: `
      *[slot='prefix'] {
        color: blue;
      }
    `
  };
}
```

`<web-component-two>`:
```js
get slotStyles() {
  return {
    prefix: `
      *[slot='prefix'] {
        color: red;
      }
    `
  };
}
```

That may be an overkill since it's always possible to define a specific enough id:
`<web-component-one>`:
```js
get slotStyles() {
  return {
    webComponentOnePrefix: `
      *[slot='prefix'] {
        color: blue;
      }
    `
  };
}
```

Reply:
Good point. I will update the logic and add a test for this specific case.